collector/xfrm.go

```diff
@@ -0,0 +1,253 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !noxfrm
+// +build !noxfrm
+
+package collector
+
+import (
+	"fmt"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs"
+)
+
+type xfrmCollector struct {
+	fs                  procfs.FS
+	inError             *prometheus.Desc
+	inBufferError       *prometheus.Desc
+	inHdrError          *prometheus.Desc
+	inNoStates          *prometheus.Desc
+	inStateProtoError   *prometheus.Desc
+	inStateModeError    *prometheus.Desc
+	inStateSeqError     *prometheus.Desc
+	inStateExpired      *prometheus.Desc
+	inStateMismatch     *prometheus.Desc
+	inStateInvalid      *prometheus.Desc
+	inTmplMismatch      *prometheus.Desc
+	inNoPols            *prometheus.Desc
+	inPolBlock          *prometheus.Desc
+	inPolError          *prometheus.Desc
+	outError            *prometheus.Desc
+	outBundleGenError   *prometheus.Desc
+	outBundleCheckError *prometheus.Desc
+	outNoStates         *prometheus.Desc
+	outStateProtoError  *prometheus.Desc
+	outStateModeError   *prometheus.Desc
+	outStateSeqError    *prometheus.Desc
+	outStateExpired     *prometheus.Desc
+	outPolBlock         *prometheus.Desc
+	outPolDead          *prometheus.Desc
+	outPolError         *prometheus.Desc
+	fwdHdrError         *prometheus.Desc
+	outStateInvalid     *prometheus.Desc
+	acquireError        *prometheus.Desc
+	logger              log.Logger
+}
+
+func init() {
+	registerCollector("xfrm", defaultEnabled, NewXfrmCollector)
+}
+
+// NewXfrmCollector returns a new Collector exposing XFRM stats.
+func NewXfrmCollector(logger log.Logger) (Collector, error) {
+	fs, err := procfs.NewFS(*procPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open procfs: %w", err)
+	}
+
+	return &xfrmCollector{
+		fs:     fs,
+		logger: logger,
+		inError: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "xfrm", "in_error_total"),
```

Comment:
These should follow Prometheus naming conventions.

```suggestion
			prometheus.BuildFQName(namespace, "xfrm", "in_errors_total"),
```

Reply:
No problem. I have added a `_packets` suffix to the metrics.

According to Linux documentation, all XFRM stats are number of packets. I hope it is more descriptive and consistent now.

> The xfrm_proc code is a set of statistics showing numbers of packets dropped by the transformation code and why.